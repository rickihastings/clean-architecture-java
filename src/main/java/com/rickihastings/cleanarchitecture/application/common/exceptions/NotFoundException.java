package com.rickihastings.cleanarchitecture.application.common.exceptions;

import an.awesome.pipelinr.AggregateException;

import java.util.List;

public class NotFoundException extends AggregateException {
    public NotFoundException(String error) {
        super(List.of(new Exception(error)));
    }
}
