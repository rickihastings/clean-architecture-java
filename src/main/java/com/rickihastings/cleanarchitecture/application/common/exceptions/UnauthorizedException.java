package com.rickihastings.cleanarchitecture.application.common.exceptions;

import an.awesome.pipelinr.AggregateException;

import java.util.List;

public class UnauthorizedException extends AggregateException {
    public UnauthorizedException() {
        super(List.of(new Exception("Unauthorized")));
    }
}
