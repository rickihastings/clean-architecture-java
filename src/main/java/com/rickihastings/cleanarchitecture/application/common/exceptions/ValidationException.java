package com.rickihastings.cleanarchitecture.application.common.exceptions;

import an.awesome.pipelinr.AggregateException;

import java.util.Collection;

public class ValidationException extends AggregateException {
    public ValidationException(Collection<Throwable> exceptions) {
        super(exceptions);
    }
}
