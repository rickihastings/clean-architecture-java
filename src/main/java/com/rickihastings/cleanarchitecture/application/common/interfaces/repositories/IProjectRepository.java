package com.rickihastings.cleanarchitecture.application.common.interfaces.repositories;

import com.rickihastings.cleanarchitecture.domain.entities.Project;
import org.springframework.stereotype.Repository;
import java.util.List;
import java.util.Optional;

@Repository
public interface IProjectRepository {
    List<Project> findAll();
    Optional<Project> findById(Long id);
    Project save(Project project);
}
