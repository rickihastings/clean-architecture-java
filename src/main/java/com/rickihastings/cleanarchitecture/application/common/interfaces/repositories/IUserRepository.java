package com.rickihastings.cleanarchitecture.application.common.interfaces.repositories;

import com.rickihastings.cleanarchitecture.domain.User;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface IUserRepository {
    Optional<User> getUserByUsername(String username);
    User save(User save);
}
