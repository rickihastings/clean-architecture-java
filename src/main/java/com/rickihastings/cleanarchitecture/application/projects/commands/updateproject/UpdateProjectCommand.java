package com.rickihastings.cleanarchitecture.application.projects.commands.updateproject;

import an.awesome.pipelinr.Command;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.rickihastings.cleanarchitecture.application.common.annotations.Authenticate;
import com.rickihastings.cleanarchitecture.application.projects.ProjectDto;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;

@Getter
@Setter
@NoArgsConstructor
@Authenticate
public class UpdateProjectCommand implements Command<ProjectDto> {
    @JsonIgnore
    private Long id;

    @JsonProperty("title")
    @NotNull(message = "title cannot be null")
    @NotBlank(message = "title cannot be blank")
    private String title;
}
