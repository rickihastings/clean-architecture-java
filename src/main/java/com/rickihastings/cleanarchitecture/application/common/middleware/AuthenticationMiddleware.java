package com.rickihastings.cleanarchitecture.application.common.middleware;

import an.awesome.pipelinr.Command;
import com.rickihastings.cleanarchitecture.application.common.annotations.Authenticate;
import com.rickihastings.cleanarchitecture.application.common.exceptions.UnauthorizedException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;

@Component
public class AuthenticationMiddleware implements Command.Middleware {

    @Override
    public <R, C extends Command<R>> R invoke(C command, Next<R> next) {
        var hasAuthenticate = command.getClass().isAnnotationPresent(Authenticate.class);

        if (hasAuthenticate) {
            var auth = SecurityContextHolder.getContext().getAuthentication();

            if (auth != null && auth.getAuthorities().stream().anyMatch(a -> a.getAuthority().equals("ROLE_USER"))) {
                return next.invoke();
            }

            throw new UnauthorizedException();
        }

        return next.invoke();
    }
}
