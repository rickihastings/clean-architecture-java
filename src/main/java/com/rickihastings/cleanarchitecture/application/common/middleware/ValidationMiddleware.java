package com.rickihastings.cleanarchitecture.application.common.middleware;

import an.awesome.pipelinr.Command;
import com.rickihastings.cleanarchitecture.application.common.exceptions.ValidationException;
import org.springframework.stereotype.Component;
import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import javax.validation.ValidatorFactory;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Set;

@Component
public class ValidationMiddleware implements Command.Middleware {

    private final Validator validator;

    public ValidationMiddleware(ValidatorFactory factory) {
        this.validator = factory.getValidator();
    }

    @Override
    public <R, C extends Command<R>> R invoke(C command, Next<R> next) {
        Set<ConstraintViolation<C>> violations = validator.validate(command);

        if (violations.size() > 0) {
            Collection<Throwable> exceptions = new ArrayList<>();
            for (ConstraintViolation<C> violation : violations) {
                exceptions.add(new Exception(violation.getMessage()));
            }

            throw new ValidationException(exceptions);
        }

        return next.invoke();
    }
}
