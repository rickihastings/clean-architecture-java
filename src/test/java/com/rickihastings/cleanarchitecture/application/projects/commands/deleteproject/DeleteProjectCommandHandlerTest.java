package com.rickihastings.cleanarchitecture.application.projects.commands.deleteproject;

import an.awesome.pipelinr.Pipelinr;
import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.rickihastings.cleanarchitecture.DefaultSnapshotSerializer;
import com.rickihastings.cleanarchitecture.application.common.exceptions.NotFoundException;
import com.rickihastings.cleanarchitecture.application.common.interfaces.repositories.IProjectRepository;
import com.rickihastings.cleanarchitecture.application.common.middleware.ValidationMiddleware;
import com.rickihastings.cleanarchitecture.application.projects.commands.updateproject.UpdateProjectCommand;
import com.rickihastings.cleanarchitecture.seeds.ProjectSeeds;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mockito;

import javax.validation.Validation;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.*;

@ExtendWith({ SnapshotExtension.class })
public class DeleteProjectCommandHandlerTest {

    private Expect expect;

    private final IProjectRepository projectRepository = Mockito.mock(IProjectRepository.class);

    private Pipelinr pipeline;

    @BeforeEach
    void setup() {
        var factory = Validation.buildDefaultValidatorFactory();

        pipeline = new Pipelinr()
                .with(() -> Stream.of(new DeleteProjectCommandHandler(projectRepository)))
                .with(() -> Stream.of(new ValidationMiddleware(factory)));
    }

    @Test
    public void shouldDeleteProjectWhenValid() {
        var id = 1L;
        var command = new DeleteProjectCommand();
        command.setId(id);
        when(projectRepository.findById(id)).thenReturn(ProjectSeeds.getProject(id));

        var result = pipeline.send(command);

        verify(projectRepository, times(1)).save(any());
        expect.serializer(DefaultSnapshotSerializer.class).toMatchSnapshot(result);
    }

    @Test
    public void shouldThrowNotFoundErrorWhenNotFound() {
        var command = new DeleteProjectCommand();
        command.setId(999L);

        assertThrows(NotFoundException.class, () -> {
            pipeline.send(command);

            verify(projectRepository, never()).save(any());
        });
    }
}
