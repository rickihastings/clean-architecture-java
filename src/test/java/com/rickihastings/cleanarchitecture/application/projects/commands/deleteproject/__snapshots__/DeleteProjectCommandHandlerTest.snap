com.rickihastings.cleanarchitecture.application.projects.commands.deleteproject.DeleteProjectCommandHandlerTest.shouldDeleteProjectWhenValid=[
  { }
]